app/controllers/incomeSources/add/CheckUKPropertyDetailsController.scala

```diff
@@ -16,60 +16,147 @@
+                                                 mcc: MessagesControllerComponents,
+                                                 val ec: ExecutionContext,
+                                                 val itvcErrorHandler: ItvcErrorHandler,
+                                                 val itvcErrorHandlerAgent: AgentItvcErrorHandler,
+                                                 val languageUtils: LanguageUtils) extends ClientConfirmedController with I18nSupport with FeatureSwitching with ImplicitDateFormatter {
+
+  def getBackUrl(isAgent: Boolean): String = {
+    if (isAgent) controllers.incomeSources.add.routes.UKPropertyAccountingMethodController.showAgent().url else
+      controllers.incomeSources.add.routes.UKPropertyAccountingMethodController.show().url
   }
 
+  def getSubmitUrl(isAgent: Boolean): Call = {
+    if (isAgent) controllers.incomeSources.add.routes.CheckUKPropertyDetailsController.submitAgent() else
+      controllers.incomeSources.add.routes.CheckUKPropertyDetailsController.submit()
+  }
 
-  def show(): Action[AnyContent] =
-    (checkSessionTimeout andThen authenticate andThen retrieveNino
-      andThen retrieveIncomeSources).async {
-      implicit user =>
-        handleRequest(
-          isAgent = false
-        )
-    }
+  def getHomePageUrl(isAgent: Boolean): Call = {
+    if (isAgent) controllers.routes.HomeController.showAgent else
+      controllers.routes.HomeController.show()
+  }
+
+  def getUKPropertyReportingMethodUrl(isAgent: Boolean, id: String): Call = {
+    if (isAgent) controllers.incomeSources.add.routes.UKPropertyReportingMethodController.showAgent(id) else
+      controllers.incomeSources.add.routes.UKPropertyReportingMethodController.show(id)
+  }
+
+  def getErrorHandler(isAgent: Boolean): FrontendErrorHandler with ShowInternalServerError = {
+    if (isAgent) itvcErrorHandlerAgent else itvcErrorHandler
+  }
+
+  def show(): Action[AnyContent] = (checkSessionTimeout andThen authenticate andThen retrieveNino
+    andThen retrieveIncomeSources andThen retrieveBtaNavBar).async {
+    implicit user =>
+      handleRequest(
+        isAgent = false
+      )
+  }
 
   def showAgent(): Action[AnyContent] = Authenticated.async {
     implicit request =>
       implicit user =>
-        getMtdItUserWithIncomeSources(incomeSourceDetailsService).flatMap {
+        getMtdItUserWithIncomeSources(incomeSourceDetailsService) flatMap {
           implicit mtdItUser =>
             handleRequest(
               isAgent = true
             )
         }
   }
 
-}
+  def handleRequest(isAgent: Boolean)(implicit user: MtdItUser[_]): Future[Result] = {
+    val backUrl = getBackUrl(isAgent)
+    val postAction = getSubmitUrl(isAgent)
+    val homePageUrl = getHomePageUrl(isAgent)
+    val errorHandler = getErrorHandler(isAgent)
+
+    if (isDisabled(IncomeSources)) {
+      Future.successful(Redirect(homePageUrl))
```

Comment:
we do this in many other places - perhaps we can pull it out into a util function?

Reply:
eg def withIcomeSourcesFS(block: => Future[Result]) = {
   if (isEnabled(incomesources)) block else {
     ...redirect
   }
 }  
 
 withIncomeSourcesFS {
   .. happy path here
 }

Reply:
good idea 👍